cmd/kube-apiserver-healthcheck/main.go

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"strings"
+
+	"k8s.io/klog"
+)
+
+// healthCheckServer is the http server
+type healthCheckServer struct {
+	tlsConfig *tls.Config
+}
+
+// handler processes a single http request
+func (s *healthCheckServer) handler(w http.ResponseWriter, r *http.Request) {
+	path := r.URL.Path
+	path = strings.TrimPrefix(path, "/")
+	tokens := strings.Split(path, "/")
+
+	if r.Method == "GET" {
+		if len(tokens) == 2 && tokens[0] == ".kube-apiserver-healthcheck" && tokens[1] == "healthz" {
+			// This is a check for our own health
+			w.WriteHeader(http.StatusOK)
+			w.Write([]byte("ok"))
+			return
+		}
+
+		if len(tokens) == 1 && tokens[0] == "readyz" {
+			s.proxyRequest(w, r)
+			return
+		}
+
+		if len(tokens) == 1 && tokens[0] == "healthz" {
+			s.proxyRequest(w, r)
+			return
+		}
+	}
+
+	klog.Infof("unknown request: %s %s", r.Method, r.URL.Path)
+	http.Error(w, "not found", http.StatusNotFound)
+}
+
+// httpClient builds an isolated http.Client
+func (s *healthCheckServer) httpClient() *http.Client {
+	transport := &http.Transport{
+		TLSClientConfig: s.tlsConfig,
+	}
```

Comment:
It seems expensive to create a new `Transport` for each incoming query. Shouldn't this be one per `healthCheckServer`?

Reply:
The fear is that they contaminate each other (e.g. cookies), but given our use case, you're right - simplifying.